polygeist/tools/cgeist/Lib/clang-mlir.cc

```diff
@@ -1213,40 +1247,51 @@ ValueCategory MLIRScanner::CommonFieldLookup(clang::QualType CT,
   }
 
   if (auto PT = dyn_cast<LLVM::LLVMPointerType>(Val.getType())) {
-    if (!isa<LLVM::LLVMStructType, LLVM::LLVMArrayType>(PT.getElementType())) {
+    if (!isa<LLVM::LLVMStructType, LLVM::LLVMArrayType>(ElementType)) {
       llvm::errs() << "Function: " << Function << "\n";
       FD->dump();
       FD->getType()->dump();
       llvm::errs() << " val: " << Val << " - pt: " << PT << " fn: " << FNum
                    << " ST: " << *ST << "\n";
     }
 
-    Type ET = TypeSwitch<Type, Type>(PT.getElementType())
+    Type ET = TypeSwitch<Type, Type>(ElementType)
                   .Case<LLVM::LLVMStructType>([FNum](LLVM::LLVMStructType ST) {
                     return ST.getBody()[FNum];
                   })
                   .Case<LLVM::LLVMArrayType, MemRefType>(
                       [](auto Ty) { return Ty.getElementType(); });
     Value CommonGep = Builder.create<LLVM::GEPOp>(
-        Loc, LLVM::LLVMPointerType::get(ET, PT.getAddressSpace()), Val,
+        Loc, Glob.getTypes().getPointerType(ET, PT.getAddressSpace()),
+        ElementType, Val,
         ValueRange({Builder.create<arith::ConstantIntOp>(Loc, 0, 32),
                     Builder.create<arith::ConstantIntOp>(Loc, FNum, 32)}),
         /* inbounds*/ true);
+    std::optional<mlir::Type> ElemTy = ET;
 
     if (RD->isUnion()) {
       LLVM::TypeFromLLVMIRTranslator TypeTranslator(*Module->getContext());
       auto SubType = TypeTranslator.translateType(
           mlirclang::getLLVMType(FD->getType(), Glob.getCGM()));
+      ElemTy = SubType;
       CommonGep = Builder.create<LLVM::BitcastOp>(
-          Loc, LLVM::LLVMPointerType::get(SubType, PT.getAddressSpace()),
+          Loc, Glob.getTypes().getPointerType(SubType, PT.getAddressSpace()),
           CommonGep);
     }
 
-    if (IsLValue)
-      CommonGep =
-          ValueCategory(CommonGep, /*isReference*/ true).getValue(Builder);
+    if (IsLValue) {
+      CommonGep = ValueCategory(CommonGep, /*isReference*/ true, ElemTy)
+                      .getValue(Builder);
+      if (auto MT = dyn_cast<MemRefType>(*ElemTy)) {
+        ElemTy = MT.getElementType();
+      } else if (auto PT = dyn_cast<LLVM::LLVMPointerType>(*ElemTy)) {
+        if (!PT.isOpaque()) {
+          ElemTy = PT.getElementType();
+        }
+      }
+    }
 
-    return ValueCategory(CommonGep, /*isReference*/ true);
+    return ValueCategory(CommonGep, /*isReference*/ true, ElemTy);
```

Comment:
Is it OK to not modify `ElemTy` in the if-then-else statement and pass that to `ValueCategory`?, e.g., for the opaque `LLVMPointerType` case